docs/2022-04-07-Full-Corectness-All-Path-Reachability-Proofs.md

```diff
@@ -0,0 +1,412 @@
+Proving Full Correctness All-Path Reachability Claims
+=====================================================
+
+This document details Full Correctness All-Path Reachability without solving the
+most-general-unifier (MGU) problem.
+MGU will be detailed in a separate document.
+
+_Prepared by Traian Șerbănuță, based on [proving All-Path Reachability
+claims](2019-03-28-All-Path-Reachability-Proofs.md)_
+
+Definitions
+-----------
+
+### All-path finally
+
+Given a formula `φ`, let `♢φ` denote the formula “all-path finally” `φ`,
+defined by:
+
+```
+♢φ := μX.φ ∨ (○X ∧ •⊤)
+```
+
+one consequence of the above is that `♢φ ≡ φ ∨ (○♢φ ∧ •⊤)`.
+
+Given this definition of all-path finally, a full correctness all-path
+reachability claim
+```
+∀x.φ(x) → ♢∃z.ψ(x,z)
+```
+basically states that if `φ(x)` holds for a configuration `γ`, for some `x`,
+then `P(γ)` holds, where `P(G)` is recursively defined on configurations as:
+* there exists `z` such that `ψ(x,z)` holds for `G`
+* or
+    * `G` is not stuck (`G → • T`) and
+    * `P(G')` for all configurations `G'` in which `G` can transition
+
+__Note:__
+Using the least fix-point (`μ`) instead of the greatest fix-point (`ν`)
+restricts paths to finite ones. Therefore, the intepretation of a reachability
+claim guarantees full-correctness, as it requires that the conclusion is
```

Comment:
Of course, your proposal is straightforward, more general, and it would be nice to have support for it in the backend eventually. However, being more general might make it also harder to prove, and the proof itself might carry a nonnegligible amount of redundancy as it disconnects proving termination from proving reachability, when in fact you will probably want to use the same invariants, etc.

The approach I presented here is inspired from Dafny, and seems to work quite well there. Also implementing it in the backend might require less effort than implementing a more general solution.

Reply:
I agree. I think it's definitely a good idea to prove both termination and correctness at the same time. What I'm suggesting is to use `reachability phi /\ WF` to justify the correctness of the "total-correctness eventually operator" encoding, by proving them equivalent using the proof system.